src/targets/gpu/include/migraphx/gpu/gemm_impl.hpp

```diff
@@ -24,28 +24,492 @@
+    {
+        if(strided_batched)
+        {
+            auto gemm_args =
+                create_strided_batched_gemm_args(ctx, input_args, int8_flag, solution_idx);
+            rocblas_invoke(&rocblas_gemm_strided_batched_ex, gemm_args);
+        }
+        else
+        {
+            auto gemm_args = create_gemm_args(ctx, input_args, int8_flag, solution_idx);
+            rocblas_invoke(&rocblas_gemm_ex, gemm_args);
+        }
+    }
+
+    auto validate(context& ctx, const std::vector<shape>& input_shapes, int32_t solution_idx) const
+    {
+        std::vector<argument> input_args;
+        std::transform(input_shapes.begin(),
+                       input_shapes.end(),
+                       std::back_inserter(input_args),
+                       [](const shape& x) { return to_gpu(generate_argument(x)); });
+
+        return validate(ctx, input_args, solution_idx);
+    }
+
+    /**
+     * Checks a particular solution for validity by running it with the flag
+     * rocblas_gemm_flags_check_solution_index, and returns the result
+     */
+    auto validate(context& ctx, const std::vector<argument>& input_args, int32_t solution_idx) const
+    {
+        if(strided_batched)
+        {
+            auto gemm_args = create_strided_batched_gemm_args(
+                ctx, input_args, rocblas_gemm_flags_check_solution_index, solution_idx);
+            return rocblas_invoke(&rocblas_gemm_strided_batched_ex, gemm_args);
+        }
+        else
+        {
+            auto gemm_args = create_gemm_args(
+                ctx, input_args, rocblas_gemm_flags_check_solution_index, solution_idx);
+            return rocblas_invoke(&rocblas_gemm_ex, gemm_args);
+        }
+    }
+
+    void print_args() const
+    {
+        std::cout << "trans: " << transa << " transb: " << transb << "\n";
+        std::cout << "m: " << m << " n: " << n << " k: " << k << "\n";
+        std::cout << "alpha: " << alpha << " beta: " << beta << "\n";
+        std::cout << "lda : " << lda << " ldb: " << ldb << " ldc: " << ldc << " ldd: " << ldd
+                  << "\n";
+        std::cout << "strided_batched: " << strided_batched << "\n";
+        std::cout << "astride: " << a_stride << " b_stride: " << b_stride
+                  << " c_stride: " << c_stride << " d_stride: " << d_stride << "\n";
+        std::cout << "arg type : " << arg_type << " compute_type: " << compute_type
+                  << " output_type: " << output_type << "\n";
+        std::cout << "int8_flag: " << int8_flag << "\n";
+    }
+
+    auto create_strided_batched_gemm_get_solutions_args(context& ctx,
+                                                        const std::vector<argument>& args,
+                                                        rocblas_int* list,
+                                                        rocblas_int* list_size) const
+    {
+        return pack(ctx.get_stream().get_rocblas(),
```

Comment:
You can refactor this using `pack_join()`which can merge two packs. 

e.g you can have function for `create_gemm_args()` which returns common args for running gemm and for calling get_solutions. 

You can then, extend gemm_args using pack_join to make it run with get_solutions.

Reply:
I don't see any examples of similar usage.  Can pack_join put the arguments in the right order?  The different API calls contain different arguments in the middle of the lists, not just concatenated at the end.  Is it possible to insert extra arguments in a specified location in the list?  (And is it even worth it, if it's that much bother?)

Reply:
pack_join would concatenate two different packs. It can't insert in the middle. 
I was saying we could keep strided_batched_gemm and regular gemm separate, but for each one separately, we could have function `create_common_gemm_args()` and `create_common_strided_batched_gemm_args()` and then we can extend them using pack_join for creating args for either running gemm or for fetching candidate solutions.

Reply:
ok, done.